### DISCUSSION TITLE
How to animate dialog (modal) with Framer Motion

### DISCUSSION THREAD

**Discussion by kwiat1990** [ariakit/ariakit#1643] [Questions] [answered]
Hi, I'm wondering how one could animate a dialog component with the help of Framer Motion. I found this discussion https://github.com/ariakit/ariakit/discussions/843, which covers a little bit of the subject and it does also work with menu component. But dialog seems to be a different story as the backdrop isn't a separate component/element but it's controlled with with `backdropProps` instead, so that the approach for popover/menu doesn't really work. At the same time taking advantae of the `backdrop` prop and passing in a `motion.div` as following cause errors:

```tsx
<Dialog
  backdrop={
    <motion.div
      key="backdrop"
      className={clsx(
        "fixed inset-0 z-50 flex justify-center bg-gray-900/75 p-4 backdrop-blur-[2px]",
        {
          "items-center": position === "center",
          "items-start": position === "top",
        }
      )}
    ></motion.div>
  }
>
...
</Dialog>
```

**Comment by diegohaz** [comment, marked as answer]
Here's an updated example with menu + framer motion: https://ariakit.org/examples/menu-framer-motion

As for the `backdrop` prop, it accepts a component, not an element, similar to the `as` prop. It would be something like this:

```jsx
<Dialog
  backdrop={motion.div}
  backdropProps={{ className: "..." }}
/>
```

**Reply by kwiat1990** [reply]
Then how props can be passed to `motion.div`, which should animate backdrop? I have managed to pass a function, which returns a `<motion.div...></div>` but then I can get the main dialog element to be displayed.

**Reply by diegohaz** [reply]
@kwiat1990 It's right there `backdropProps`.

The `backdrop` behaves similarly to the `as` prop. You can't pass an element (`<motion.div />`) or an inline function (`() => <motion.div />`).

**Reply by diegohaz** [reply]
Alternatively, you can create a separate backdrop component:
```jsx
const MyBackdrop = forwardRef((props, ref) => {
  return <motion.div {...props} ref={ref} myCustomProp />
});

function MyDialog() {
  return <Dialog backdrop={MyBackdrop} />;
}
```

**Reply by kwiat1990** [reply]
Well, passing all values via `backdropProps` would be great, but as soon as I define `backdrop={motion.div}` TS starts to complain. Using `backdropProps` with values for the animation triggers additionals TS issues.

```
No overload matches this call.
  Overload 2 of 2, '(props: Props<DialogOptions<"div">>): Element | null', gave the following error.
    Type 'ForwardRefComponent<HTMLDivElement, HTMLMotionProps<"div">>' is not assignable to type 'boolean | ElementType<Pick<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, keyof HTMLAttributes<HTMLDivElement> | "key"> & { ...; }> | undefined'.
  Overload 2 of 2, '(props: Props<DialogOptions<"div">>): Element | null', gave the following error.
    Type '{ initial: string; animate: string; exit: string; }' is not assignable to type 'Omit<DisclosureContentProps<"div">, "state">'.
      Object literal may only specify known properties, and 'initial' does not exist in type 'Omit<DisclosureContentProps<"div">, "state">'.
```

**Reply by diegohaz** [reply]
Yeah, you can either ignore the TS errors (with `// @ts-ignore` or casting `as any`) or create a separate component, as I showed above.

**Reply by diegohaz** [reply]
We may introduce a `renderBackdrop` prop so you can pass an inline function. Feel free to submit a feature request if you need this.

**Reply by kwiat1990** [reply]
@diegohaz thanks for support and some insights. I've ended up with a separate component for the backdrop.
